From 17a7c8a1da8a13cc1e8374d3624e4c6b0a2fffd3 Mon Sep 17 00:00:00 2001
From: chenxiaolong <chenxiaolong@loongson.cn>
Date: Tue, 30 Jan 2024 09:20:40 +0800
Subject: [PATCH v58] asan.c (asan_emit_stack_protection): Use full-sized mask
 to align the base address on 64-bit strict-alignment platforms.

commit 362432c00db860483058ff609a893151bf8e4b1c
Author: Eric Botcazou <ebotcazou@gcc.gnu.org>
Date:   Fri Feb 15 21:40:24 2019 +0000

    asan.c (asan_emit_stack_protection): Use full-sized mask to align the base address on 64-bit strict-alignment platforms.

            * asan.c (asan_emit_stack_protection): Use full-sized mask to align
            the base address on 64-bit strict-alignment platforms.

    From-SVN: r268949

Change-Id: I91b37f65b0c3c68aab2106f45c4b25a9f57d5d06
---
 src/gcc/asan.c | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/src/gcc/asan.c b/src/gcc/asan.c
index 235e21947..23c1fc749 100644
--- a/src/gcc/asan.c
+++ b/src/gcc/asan.c
@@ -1309,11 +1309,13 @@ asan_emit_stack_protection (rtx base, rtx pbase, unsigned int alignb,
     }
   /* Align base if target is STRICT_ALIGNMENT.  */
   if (STRICT_ALIGNMENT)
-    base = expand_binop (Pmode, and_optab, base,
-			 gen_int_mode (-((GET_MODE_ALIGNMENT (SImode)
-					  << ASAN_SHADOW_SHIFT)
-					 / BITS_PER_UNIT), Pmode), NULL_RTX,
-			 1, OPTAB_DIRECT);
+    {
+      const HOST_WIDE_INT align
+       = (GET_MODE_ALIGNMENT (SImode) / BITS_PER_UNIT) << ASAN_SHADOW_SHIFT;
+      base = expand_binop (Pmode, and_optab, base, gen_int_mode (-align, Pmode),
+                          NULL_RTX, 1, OPTAB_DIRECT);
+    }
+
 
   if (use_after_return_class == -1 && pbase)
     emit_move_insn (pbase, base);
-- 
2.20.1

